### PR TITLE
gpiobus(4): Add an acpi variant of gpiobus

### PR DESCRIPTION
**Pull request by VexedUXR**
Add a simple acpi_gpiobus driver

Currently, this only installs an address space handler and attempts to set pin flags using info obtained from GPIO resources. Child enumeration doesn't seem to be necessary at this point, as it would only complicate things. It can be added in the future if needed.

Tested on an Acer one 10 s1002

### CODE REVIEW

**Comment by VexedUXR** [comment]
CC @wulf7

File: sys/dev/gpio/acpi_gpiobus.c [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+static int
+acpi_gpiobus_attach(device_t dev)
+{
+	struct acpi_gpiobus_softc *sc;
+	struct acpi_gpiobus_ctx ctx;
+	ACPI_HANDLE handle;
+	ACPI_STATUS status;
+	int err;
+
+	if ((err = gpiobus_init_softc(dev)) != 0)
```
Comment (wulf7) [inline sys/dev/gpio/acpi_gpiobus.c:232]: Use attach method from base driver

File: sys/dev/gpio/acpi_gpiobus.c [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+	if (res->Type != ACPI_RESOURCE_TYPE_GPIO)
+		return (AE_OK);
+
+	if (ACPI_FAILURE(AcpiGetHandle(ACPI_ROOT_OBJECT,
+	    gpio_res->ResourceSource.StringPtr, &handle)) ||
+	    handle != ctx->dev_handle)
+		return (AE_OK);
+
+	/* Figure out pin flags */
+	flags = 0;
```
Comment (wulf7) [inline sys/dev/gpio/acpi_gpiobus.c:67]: acpi_resource to gpio flags conversion can be factored out to separate function

File: sys/dev/gpio/acpi_gpiobus.c [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+		}
+
+	switch (gpio_res->PinConfig) {
+	case ACPI_PIN_CONFIG_PULLUP:
+		flags |= GPIO_PIN_PULLUP;
+		break;
+	case ACPI_PIN_CONFIG_PULLDOWN:
+		flags |= GPIO_PIN_PULLDOWN;
+		break;
+	}
```
Comment (wulf7) [inline sys/dev/gpio/acpi_gpiobus.c:100]: There is also gpio_res->Shareable field to process. It should be mapped to new GPIO_INTR_SHARED flag in sys/gpio.h

File: sys/dev/gpio/acpi_gpiobus.c [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+		UINT16 pin = gpio_res->PinTable[i];
+
+		if (__predict_false(pin >= super_sc->sc_npins)) {
+			device_printf(super_sc->sc_busdev,
+			    "invalid pin 0x%x, max: 0x%x (bad ACPI tables?)\n",
+			    pin, super_sc->sc_npins - 1);
+			return (AE_LIMIT);
+		}
+
+		GPIO_PIN_SETFLAGS(super_sc->sc_dev, pin, flags);
```
Comment (wulf7) [inline sys/dev/gpio/acpi_gpiobus.c:119]: GPIO_PIN_SETFLAGS sets only pin config and IO direction. Please filter out interupt bits with logical AND

File: sys/dev/gpio/acpi_gpiobus.c (hunk 3 above)
Comment (VexedUXR) [reply sys/dev/gpio/acpi_gpiobus.c:100]: Oh, nice catch.

Would we need to change the interrupt handling code in gpioc to accommodate the flag?
But then again, I think this attribute only means "Can this be accessed/is this used by other ACPI devices". So in that case, it shouldn't affect interrupt creation.
Hm, I'm not sure.

**Comment by bsdimp** [comment]
@wulf7 are you happy with this...

**Comment by wulf7** [comment]
> @wulf7 are you happy with this...

Almost. I can handle remaining very minor issues after returning of vacations

**Comment by wulf7** [comment]
Committed. Thanks!

I ifdef-ed out interrupt-related parts as they are not used yet and requires some extra work in gpiobus.c and gpioc.c

**Comment by VexedUXR** [comment]
Thanks!

**Pull request closed without merging**